pkg/gds/store/wire/wire.go

```diff
@@ -0,0 +1,204 @@
+*/
+package wire
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/trisacrypto/directory/pkg/gds/global/v1"
+	"github.com/trisacrypto/directory/pkg/gds/models/v1"
+	"github.com/trisacrypto/directory/pkg/gds/peers/v1"
+	"github.com/trisacrypto/directory/pkg/gds/store"
+	pb "github.com/trisacrypto/trisa/pkg/trisa/gds/models/v1beta1"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+var ErrCannotReplicate = errors.New("object in namespace cannot be replicated")
+
+// UnmarshalProto expects protocol buffer data and unmarshals it to the correct type
+// based on the namespace. This is a utility function for dealing with the various
+// namespaces and types that GDS manages and is not a substitute for direct unmarshaling.
+func UnmarshalProto(namespace string, data []byte) (_ proto.Message, err error) {
+	switch namespace {
+	case store.NamespaceVASPs:
+		vasp := &pb.VASP{}
+		if err = proto.Unmarshal(data, vasp); err != nil {
+			return nil, fmt.Errorf("could not unmarshal %s to %T: %s", namespace, vasp, err)
+		}
+		return vasp, nil
+	case store.NamespaceCertReqs:
+		certreq := &models.CertificateRequest{}
+		if err = proto.Unmarshal(data, certreq); err != nil {
+			return nil, fmt.Errorf("could not unmarshal %s to %T: %s", namespace, certreq, err)
+		}
+		return certreq, nil
+	case store.NamespaceReplicas:
+		peer := &peers.Peer{}
+		if err = proto.Unmarshal(data, peer); err != nil {
+			return nil, fmt.Errorf("could not unmarshal %s to %T: %s", namespace, peer, err)
+		}
+		return peer, nil
+	default:
+		return nil, fmt.Errorf("unknown namespaces %q", namespace)
+	}
+}
+
+// UnmarshalObject expects protocol buffer data and unmarshals it to a *global.Object,
+// including the original data as a marshaled anypb.Any on the Object data.
+func UnmarshalObject(namespace string, data []byte) (obj *global.Object, err error) {
+	switch namespace {
+	case store.NamespaceVASPs:
+		// Unmarshal the VASP
+		vasp := &pb.VASP{}
+		if err = proto.Unmarshal(data, vasp); err != nil {
+			return nil, fmt.Errorf("could not unmarshal %s to %T: %s", namespace, vasp, err)
+		}
+
+		// Fetch the object metadata for the VASP
+		if obj, _, err = models.GetMetadata(vasp); err != nil {
+			return nil, err
+		}
+
+		// Marshal the VASP data back onto the Any field of the object
+		if obj.Data, err = anypb.New(vasp); err != nil {
+			return nil, err
+		}
+	case store.NamespaceCertReqs:
+		certreq := &models.CertificateRequest{}
+		if err = proto.Unmarshal(data, certreq); err != nil {
+			return nil, fmt.Errorf("could not unmarshal %s to %T: %s", namespace, certreq, err)
+		}
+
+		// Fetch the object metadata and add the CertificateRequest data back onto the Any field
+		obj = certreq.Metadata
+		if obj.Data, err = anypb.New(certreq); err != nil {
+			return nil, err
+		}
+	case store.NamespaceReplicas:
```

Comment:
Because this handles NamespaceReplicas, I believe that gdsutil will replicate peers as a result.